# Current Topics in Bioinformatics workshops

| Audience | Computational Skills | Duration |
:----------|:----------|:----------|
| Biologists | Beginner or intermediate R and/or beginner bash | 2-3 hour workshops |

<p align="center">
Use https://tinyurl.com/hcbc-modules to access this page.
</p>

This repository has teaching materials for **2-3 hour**, hands-on workshops covering a variety of topics related to bioinformatics data analysis. The workshops will lead participants through performing different types of analyses using R/RStudio or Linux. 

Some workshops will require a working knowledge of R or completion of the [Introduction to R workshop](IntroR). Other workshops will require a working knowledge of the bash scripting language or completion of the [Introduction to Shell workshop](Intro_shell/). 

** **NOTE: Detailed information** and **preparation instructions** for each of the workshops can be found by clicking on the workshop links in the table below.

> Workshop introduction slides are available [here](https://github.com/hbctraining/Training-modules/raw/master/Intro_current_topics_online_2022.pdf).

***

### Summer-Fall 2022 Schedule (1pm - 4pm):

| Topic and Link(s) to lessons | Prerequisites | Date | Registration |
|:---------------|:-------------:|:-------------:|:-------------:|
| [Basics of Python](Python) | None | 8/10/2022	| [Sign up!](https://harvard.zoom.us/meeting/register/tJMvd-2gqjMtHNN70z8--PgwRfXmip3FwGza) |
| [R practice exercises]() | Beginner R or [Online R course - Harvard Catalyst](https://catalyst.harvard.edu/courses/intro-to-r/) | 9/7/2022	| [Sign up!](https://harvard.zoom.us/meeting/register/tJ0vce6tpjoqGNBRUO-ha0mWW2RYv6xV5exd) |
| [Generating reports with Rmarkdown]() | Beginner R or [Online R course - Harvard Catalyst](https://catalyst.harvard.edu/courses/intro-to-r/) | 10/5/2022	| [Sign up!](https://harvard.zoom.us/meeting/register/tJEpde-prT4iH9wwbzxxqlBiuW_4Ovajt7HC) |
| [Introduction to Shell](https://hbctraining.github.io/Training-modules/Intro_shell/)	| None | 11/2/2022	| [Sign up!](https://harvard.zoom.us/meeting/register/tJEuf--hpz0pGd3_Wu0vBE1mcxSMaSb88KUB) |
| [Version control using Git/Github]() | [Introduction to Shell]() | 12/7/2022	| [Sign up!](https://harvard.zoom.us/meeting/register/tJ0kdeCtqz8sGtJx6DCaXw9lqNGkFJ5SbP3G) | 

***

## Links to workshop materials

### R-based workshops:

| Topic and Link(s) to lessons | Prerequisites |
|:---------------|:-------------:|
| [Introduction to R](IntroR) | None |
| [Tidyverse](Tidyverse_ggplot2) | Beginner R or [Intro to R workshop](IntroR) |
| [Introduction to R Practical](IntroR_practical_online_resource) | Beginner R or [Completion of the Intro to R online resource](https://projects.iq.harvard.edu/hcatrresource/) |
| [Gene annotations and functional analysis of gene lists](DGE-functional-analysis) | Beginner R or [Intro to R workshop](IntroR) |
| [Generating research analysis reports with RMarkdown](Rmarkdown) | Beginner R or [Intro to R workshop](IntroR) |
| [Interactive Data Visualization with Shiny in R (with Ista Zahn from the Harvard Business School)](https://github.com/izahn/shiny_workshop) | Beginner R or [Intro to R workshop](IntroR) |
| [Publication Perfect: Part I](publication_perfect#part-i) | Beginner R or [Completion of the Intro to R online resource](https://projects.iq.harvard.edu/hcatrresource/) |
| [Publication Perfect: Part II](publication_perfect#part-ii) | [Publication Perfect: Part I](publication_perfect#part-i) |
| [Functional analysis of gene lists](DGE-functional-analysis/) | Beginner R or [Intro to R workshop](IntroR) |

### Shell-based workshops:

| Topic and Link(s) to lessons | Prerequisites |
|:---------------|:-------------:|
| [Introduction to the command-line interface]Intro_shell) | None |
| [Intermediate bash](Intermediate_shell) | Beginner shell or [Intro to the command-line interface](Intro_shell) |
| [Version control using Git and Github](Git-Github) | Beginner shell or [Intro to the command-line interface](Intro_shell)  |
| [Accessing genomic reference and experimental sequencing data](https://hbctraining.github.io/Accessing_public_genomic_data) | Beginner shell or [Intro to the command-line interface](Intro_shell)  |
| [Exploring genomic variants using GEMINI](Exploring_variants_with_GEMINI) | Beginner shell or [Intro to the command-line interface](Intro_shell) |

### Additional workshops:

| Topic and Link(s) to lessons | Prerequisites |
|:---------------|:-------------:|
| [Introduction to Python](Python) | None |
| [Planning a bulk RNA-seq analysis: Part I](planning_successful_rnaseq#part-i) | None |
| [Planning a bulk RNA-seq analysis: Part II](planning_successful_rnaseq#part-ii) | None |
| [Make your (RNA-seq) data analysis reproducible](reproducible_analyses)- *Taught by [Julie Goldman](https://scholar.harvard.edu/julie_goldman) from Countway Library* | None |
| [Improving your (RNA-seq) data analysis using version control (Git)](https://hbctraining.github.io/versioning_data_scripts/) | None |


***

*These materials have been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*
